From patchwork Sun Oct  7 14:44:25 2018
Content-Type: text/plain; charset="utf-8"
MIME-Version: 1.0
Content-Transfer-Encoding: 7bit
X-Patchwork-Submitter: Wenwen Wang <wang6495@umn.edu>
X-Patchwork-Id: 10629663
Return-Path: <kvm-owner@kernel.org>
Received: from mail.wl.linuxfoundation.org (pdx-wl-mail.web.codeaurora.org
 [172.30.200.125])
	by pdx-korg-patchwork-2.web.codeaurora.org (Postfix) with ESMTP id BD49E112B
	for <patchwork-kvm@patchwork.kernel.org>;
 Sun,  7 Oct 2018 14:54:04 +0000 (UTC)
Received: from mail.wl.linuxfoundation.org (localhost [127.0.0.1])
	by mail.wl.linuxfoundation.org (Postfix) with ESMTP id 9EAEC28DD8
	for <patchwork-kvm@patchwork.kernel.org>;
 Sun,  7 Oct 2018 14:54:04 +0000 (UTC)
Received: by mail.wl.linuxfoundation.org (Postfix, from userid 486)
	id 919B328F67; Sun,  7 Oct 2018 14:54:04 +0000 (UTC)
X-Spam-Checker-Version: SpamAssassin 3.3.1 (2010-03-16) on
	pdx-wl-mail.web.codeaurora.org
X-Spam-Level: 
X-Spam-Status: No, score=-8.0 required=2.0 tests=BAYES_00,DKIM_SIGNED,
	DKIM_VALID,DKIM_VALID_AU,MAILING_LIST_MULTI,RCVD_IN_DNSWL_HI autolearn=ham
	version=3.3.1
Received: from vger.kernel.org (vger.kernel.org [209.132.180.67])
	by mail.wl.linuxfoundation.org (Postfix) with ESMTP id 3293E28DD8
	for <patchwork-kvm@patchwork.kernel.org>;
 Sun,  7 Oct 2018 14:54:04 +0000 (UTC)
Received: (majordomo@vger.kernel.org) by vger.kernel.org via listexpand
        id S1728129AbeJGWBb (ORCPT
        <rfc822;patchwork-kvm@patchwork.kernel.org>);
        Sun, 7 Oct 2018 18:01:31 -0400
Received: from mta-p7.oit.umn.edu ([134.84.196.207]:47858 "EHLO
        mta-p7.oit.umn.edu" rhost-flags-OK-OK-OK-OK) by vger.kernel.org
        with ESMTP id S1726527AbeJGWBb (ORCPT <rfc822;kvm@vger.kernel.org>);
        Sun, 7 Oct 2018 18:01:31 -0400
X-Greylist: delayed 563 seconds by postgrey-1.27 at vger.kernel.org;
 Sun, 07 Oct 2018 18:01:30 EDT
Received: from localhost (unknown [127.0.0.1])
        by mta-p7.oit.umn.edu (Postfix) with ESMTP id F41B81020
        for <kvm@vger.kernel.org>; Sun,  7 Oct 2018 14:44:39 +0000 (UTC)
X-Virus-Scanned: amavisd-new at umn.edu
Received: from mta-p7.oit.umn.edu ([127.0.0.1])
        by localhost (mta-p7.oit.umn.edu [127.0.0.1]) (amavisd-new,
 port 10024)
        with ESMTP id UU4TYpENzEjt for <kvm@vger.kernel.org>;
        Sun,  7 Oct 2018 09:44:39 -0500 (CDT)
Received: from mail-io1-f72.google.com (mail-io1-f72.google.com
 [209.85.166.72])
        (using TLSv1.2 with cipher AES128-GCM-SHA256 (128/128 bits))
        (No client certificate requested)
        by mta-p7.oit.umn.edu (Postfix) with ESMTPS id CB7E2100D
        for <kvm@vger.kernel.org>; Sun,  7 Oct 2018 09:44:39 -0500 (CDT)
Received: by mail-io1-f72.google.com with SMTP id i20-v6so17350651ioh.4
        for <kvm@vger.kernel.org>; Sun, 07 Oct 2018 07:44:39 -0700 (PDT)
DKIM-Signature: v=1; a=rsa-sha256; c=relaxed/relaxed;
        d=umn.edu; s=google;
        h=from:to:cc:subject:date:message-id;
        bh=zmJW304Drts08w59yTcGYvwTGz4bjKkqz7qCSzjGoZ4=;
        b=NDspYlSousdaNdoya8xkYzkI2eVao1E+qLaUwNfbz7IT9Ey8uxRn+iZghYvcYA3JIY
         3l/3SRHgTKhRA4FzcAfbcXVbc1N/jyQ3b8V/VL3dPhcJeXNhL9SFNBT5C3SjhY81i61X
         jhh4iaFC7AuHbEywlXdyNhYZyfDlgeFROE9wmnJWzo2Mcse/WmeNSXDuij4EhJ3XEdNl
         VK1mOgVCZDyeW7vF2TXgFr7EgkYichTcCPMbVWBcaFa+393x99JpLiZ9Pkx8c1SXUuNJ
         URsM2RidfIG50P0Y9qVmksTVCe0FCM9sEW010y5sOEP3VUq3kJn6oZS7AMSS8rvewa2g
         e0eg==
X-Google-DKIM-Signature: v=1; a=rsa-sha256; c=relaxed/relaxed;
        d=1e100.net; s=20161025;
        h=x-gm-message-state:from:to:cc:subject:date:message-id;
        bh=zmJW304Drts08w59yTcGYvwTGz4bjKkqz7qCSzjGoZ4=;
        b=raboPzCdPNJFIozr8JecJZ3PGBFzY4WVX/nIDd9M2YBND1M/XBTWsw986UVUN2ZSRd
         nkFgUpdz/rjfRuk3Jt4ONPIVjTDrXE/xDHEhB8+v4lbxJPw5dB9Tlq+wpDh3f56aCPAi
         Nw78UBFnW/45YiomgA0w7UOKMGfwwp6rF1Uyo+Rs2Q7yS/yTZuaW3hk+d8sGrUmmgElN
         g+iBpEhLb7TcSfRRguqeiGwD1Ja5exmoBUNGRjuiSoR4pWUZarLM5yhkJR4gEJPQ7af/
         xXOCeyd00ABNFxOQnfltRWLTivIFDcq0nr+NxEclv1luQ+fa1XdJ7lkMiHz8AJlFLix6
         LGCw==
X-Gm-Message-State: ABuFfoidd9ihofn9eIDa71l3IvJkTACEi6DbS+7JLp7a73FSBZQGLkf2
        zKQLOO76gY1m8VI9lqeehnFvckD/0g88h9RZzR3ertVB/zg5y7DU6Zo1gZcWG1NgI651akVbAld
        LuKqG4Ywpn2DBwRph
X-Received: by 2002:a6b:1f8b:: with SMTP id
 f133-v6mr13315508iof.179.1538923479406;
        Sun, 07 Oct 2018 07:44:39 -0700 (PDT)
X-Google-Smtp-Source: 
 ACcGV60pBuhba1ijY7WQ3k77P1Z0AQjKPMPSzgC/odPaSIeoidnTvjyGvpkVoGygVQERjkhbbY7SFg==
X-Received: by 2002:a6b:1f8b:: with SMTP id
 f133-v6mr13315500iof.179.1538923479199;
        Sun, 07 Oct 2018 07:44:39 -0700 (PDT)
Received: from cs-u-cslp16.cs.umn.edu (cs-u-cslp16.cs.umn.edu.
 [134.84.121.95])
        by smtp.gmail.com with ESMTPSA id
 n5-v6sm2115015ioh.58.2018.10.07.07.44.37
        (version=TLS1_2 cipher=ECDHE-RSA-AES128-SHA bits=128/128);
        Sun, 07 Oct 2018 07:44:38 -0700 (PDT)
From: Wenwen Wang <wang6495@umn.edu>
To: Wenwen Wang <wang6495@umn.edu>
Cc: Kangjie Lu <kjlu@umn.edu>,
        Alex Williamson <alex.williamson@redhat.com>,
        kvm@vger.kernel.org (open list:VFIO DRIVER),
        linux-kernel@vger.kernel.org (open list)
Subject: [PATCH] drivers/vfio: Fix a redundant copy bug
Date: Sun,  7 Oct 2018 09:44:25 -0500
Message-Id: <1538923466-29705-1-git-send-email-wang6495@umn.edu>
X-Mailer: git-send-email 2.7.4
Sender: kvm-owner@vger.kernel.org
Precedence: bulk
List-ID: <kvm.vger.kernel.org>
X-Mailing-List: kvm@vger.kernel.org
X-Virus-Scanned: ClamAV using ClamSMTP

In vfio_spapr_iommu_eeh_ioctl(), if the ioctl command is VFIO_EEH_PE_OP,
the user-space buffer 'arg' is copied to the kernel object 'op' and the
'argsz' and 'flags' fields of 'op' are checked. If the check fails, an
error code EINVAL is returned. Otherwise, 'op.op' is further checked
through a switch statement to invoke related handlers. If 'op.op' is
VFIO_EEH_PE_INJECT_ERR, the whole user-space buffer 'arg' is copied again
to 'op' to obtain the err information. However, in the following execution
of this case, the fields of 'op', except the field 'err', are actually not
used. That is, the second copy has a redundant part. Therefore, for both
performance and security reasons, the redundant part of the second copy
should be removed.

This patch removes such a part in the second copy. It only copies the 'err'
information from the buffer 'arg'.

Signed-off-by: Wenwen Wang <wang6495@umn.edu>
---
 drivers/vfio/vfio_spapr_eeh.c | 6 +++---
 1 file changed, 3 insertions(+), 3 deletions(-)

diff --git a/drivers/vfio/vfio_spapr_eeh.c b/drivers/vfio/vfio_spapr_eeh.c
index 38edeb4..5bc4b60 100644
--- a/drivers/vfio/vfio_spapr_eeh.c
+++ b/drivers/vfio/vfio_spapr_eeh.c
@@ -86,10 +86,10 @@ long vfio_spapr_iommu_eeh_ioctl(struct iommu_group *group,
 			ret = eeh_pe_configure(pe);
 			break;
 		case VFIO_EEH_PE_INJECT_ERR:
-			minsz = offsetofend(struct vfio_eeh_pe_op, err.mask);
-			if (op.argsz < minsz)
+			if (op.argsz < sizeof(op))
 				return -EINVAL;
-			if (copy_from_user(&op, (void __user *)arg, minsz))
+			if (copy_from_user(&op.err, (char __user *)arg +
+						minsz, sizeof(op.err)))
 				return -EFAULT;
 
 			ret = eeh_pe_inject_err(pe, op.err.type, op.err.func,
